# Defining global variables

In [1]:
REPO_NAME = 'Textual_Analysis_in_Finance'
BASE_DIR = f'/kaggle/working/{REPO_NAME}'
WEEK = 3

In [2]:
import warnings
warnings.filterwarnings('ignore')

# Clone the lecture's git repo

In [3]:
!git clone https://github.com/minhtriphan/{REPO_NAME}.git
%cd {REPO_NAME}

Cloning into 'Textual_Analysis_in_Finance'...
remote: Enumerating objects: 201, done.
remote: Counting objects: 100% (201/201), done.
remote: Compressing objects: 100% (155/155), done.
remote: Total 201 (delta 61), reused 153 (delta 30), pack-reused 0 (from 0)
Receiving objects: 100% (201/201), 7.71 MiB | 17.97 MiB/s, done.
Resolving deltas: 100% (61/61), done.
/kaggle/working/Textual_Analysis_in_Finance


# Exploring the Google pre-trained Word2Vec model

Many organizations/people train their Word2Vec models and published them. Those models are called _pre-trained models_.

One of the most popular pre-trained models is the Word2Vec model trained by Google using their Google News dataset.

The model is trained on the data containing approximately 100 billion words, generated from 3 million unique words (i.e., the vocab size). Each word is represented by a 300-dimensional vector.

We can obtain the model using `gensim`. However, first, we need to add the model as our notebook input.

The model can be found [**here**](https://www.kaggle.com/datasets/sugataghosh/google-word2vec). To use it, follow the below steps:

1. On the right-hand side tab, click on `+ Add Input` button in the `Input` tab
2. On the search bar, input the URL: https://www.kaggle.com/datasets/sugataghosh/google-word2vec
3. The model is uploaded in the first search result, named `Google Word2Vec`
4. You will see a plus sign with a circle symbol next to the search result, click on it, and wait
5. Kaggle will inform you when the model is added successfully

To use the model, we need to import the `KeyedVectors` class from `gensim`

In [4]:
from gensim.models import KeyedVectors

GOOGLE_WORD2VEC_PATH = '/kaggle/input/datasets/sugataghosh/google-word2vec/GoogleNews-vectors-negative300.bin'
google_wv = KeyedVectors.load_word2vec_format(GOOGLE_WORD2VEC_PATH, binary = True)

#### Check the vocabulary size

In [5]:
len(google_wv)

3000000

#### Check the vector size/dimensions

In [6]:
google_wv.vector_size

300

#### What are words in the vocabulary

In [7]:
list(google_wv.key_to_index.keys())[:10]    # Just print the first 10 words in the vocabulary

['</s>', 'in', 'for', 'that', 'is', 'on', '##', 'The', 'with', 'said']

#### Let check the vector of a word

In [8]:
google_wv['profit']

array([-1.49414062e-01, -1.19628906e-01, -1.47460938e-01,  1.68945312e-01,
       -2.14843750e-01, -3.45703125e-01,  1.55029297e-02, -5.54199219e-02,
        2.92968750e-01,  1.41601562e-01, -2.63671875e-01,  1.38671875e-01,
        1.77734375e-01,  1.06933594e-01, -3.28125000e-01,  1.15234375e-01,
       -1.17675781e-01, -8.48388672e-03,  3.02734375e-01, -4.51660156e-02,
        3.71093750e-02,  6.98242188e-02, -8.39843750e-02, -2.57812500e-01,
        9.86328125e-02,  1.58203125e-01, -1.05468750e-01,  3.56445312e-02,
       -1.22558594e-01, -2.37304688e-01, -7.37304688e-02,  1.36108398e-02,
       -2.13867188e-01, -1.01074219e-01, -3.93066406e-02,  1.51367188e-01,
       -1.93359375e-01,  3.73535156e-02, -2.70996094e-02,  2.77343750e-01,
        2.77343750e-01, -2.27355957e-03,  2.18750000e-01,  8.30078125e-02,
       -5.68847656e-02, -3.24218750e-01,  4.00390625e-02,  2.67578125e-01,
       -2.61718750e-01,  1.91406250e-01,  2.71484375e-01,  5.64575195e-03,
       -7.91015625e-02,  

#### Let's check the most similar words of a word

In [9]:
google_wv.most_similar('profit', topn = 5)

[('profits', 0.802002489566803),
 ('proft', 0.7426661252975464),
 ('pretax_profit', 0.6691470742225647),
 ('pretax_profits', 0.643925130367279),
 ('Profit', 0.6339423060417175)]

#### We can also compute the similarity between 2 words

In [10]:
google_wv.similarity('loss', 'profit')

np.float32(0.34199455)

#### Check which word is most different from the others

In [11]:
google_wv.doesnt_match(['stock', 'volume', 'returns', 'liquidity', 'volleyball', 'market'])

'volleyball'

#### Derive the vector for the whole sentence

Let's say we have a sentence (after cleaning): `income increase`, we can find the vector of this sentence by averaging the vectors of component words. Technically, if

* `income` $\rightarrow$ $v_1$
* `increase` $\rightarrow$ $v_2$

`income increase` $\rightarrow$ $\frac{1}{2}(v_1 + v_2)$

In [12]:
import numpy as np

sentence = 'income increase'

sentence_vector = []

for word in sentence.split():
    sentence_vector.append(google_wv[word])
    
sentence_vector = np.mean(sentence_vector, axis = 0)
sentence_vector

array([ 0.04858398, -0.11425781, -0.18554688,  0.1116333 , -0.05456543,
       -0.1274414 , -0.00976562, -0.17041016,  0.15576172,  0.10290527,
       -0.05328369,  0.08605957, -0.00817871,  0.1385498 , -0.10205078,
        0.03076172, -0.01367188,  0.01147461,  0.08251953,  0.04048157,
       -0.10400391, -0.0246582 ,  0.01342773, -0.23242188, -0.10910034,
        0.01928711, -0.01901245,  0.08728027, -0.17675781,  0.07421875,
        0.02886963, -0.2998047 ,  0.10449219, -0.19140625,  0.00146484,
       -0.05004883, -0.02111816,  0.02294922, -0.08007812,  0.29882812,
        0.1772461 ,  0.19970703,  0.14892578, -0.0065918 , -0.25976562,
       -0.36816406,  0.07009888,  0.18701172, -0.20629883,  0.05224609,
        0.15966797, -0.17333984, -0.0748291 , -0.09594727,  0.0604248 ,
       -0.07678223, -0.21533203, -0.00244141,  0.01184082, -0.07794189,
        0.05712891, -0.03617859, -0.26660156, -0.02450562,  0.23876953,
       -0.02148438, -0.15344238, -0.04730225,  0.0456543 ,  0.06

# Training a Word2Vec model using Tesla's 10-K files from 2020 to 2025

In [13]:
import os
from glob import glob

all_10k_files = glob(os.path.join(BASE_DIR, 'Data', 'FinancialStatements', '*.txt'))
all_10k_files

['/kaggle/working/Textual_Analysis_in_Finance/Data/FinancialStatements/20230131_10-K_edgar_data_1318605_0000950170-23-001409.txt',
 '/kaggle/working/Textual_Analysis_in_Finance/Data/FinancialStatements/20180223_10-K_edgar_data_1318605_0001564590-18-002956.txt',
 '/kaggle/working/Textual_Analysis_in_Finance/Data/FinancialStatements/20200213_10-K_edgar_data_1318605_0001564590-20-004475.txt',
 '/kaggle/working/Textual_Analysis_in_Finance/Data/FinancialStatements/20240129_10-K_edgar_data_1318605_0001628280-24-002390.txt',
 '/kaggle/working/Textual_Analysis_in_Finance/Data/FinancialStatements/20220207_10-K_edgar_data_1318605_0000950170-22-000796.txt',
 '/kaggle/working/Textual_Analysis_in_Finance/Data/FinancialStatements/20210208_10-K_edgar_data_1318605_0001564590-21-004599.txt',
 '/kaggle/working/Textual_Analysis_in_Finance/Data/FinancialStatements/20250130_10-K_edgar_data_1318605_0001628280-25-003063.txt',
 '/kaggle/working/Textual_Analysis_in_Finance/Data/FinancialStatements/20160224_10-

#### <span style="color:red">Questions</span>

1. Can you explain the file name? E.g., `20250130_10-K_edgar_data_1318605_0001628280-25-003063.txt`

#### <span style="color:red">Answer</span>

In the file `20250130_10-K_edgar_data_1318605_0001628280-25-003063.txt`

* `20250130`: the filing date in the `YYYYmmdd` format; the date the file is successfully submitted to SEC's system
* `10-K`: form type
* `edgar_data`: the Electronic Data Gathering, Analysis, and Retrieval system; [**details**](https://en.wikipedia.org/wiki/Electronic_Data_Gathering,_Analysis,_and_Retrieval)
* `1318605`: the Central Index Key (CIK); the unique number, i.e., company identifier, assigned by the SEC to individuals, corporations, and entities that submit filings to the EDGAR database
* `0001628280-25-003063`: the unique ID for the filing, a.k.a the accession number. It has structure too:
  * `0001628280`: the submitter/filer ID
  * `25`: year
  * `003063`: a sequential counter

#### Let's try to check one 10-K document

In [14]:
with open(all_10k_files[0], 'r') as f:
    content = f.read()
content[:5_000]

'<Header>\n<FileStats>\n    <FileName>20230131_10-K_edgar_data_1318605_0000950170-23-001409.txt</FileName>\n    <GrossFileSize>31445171</GrossFileSize>\n    <NetFileSize>807949</NetFileSize>\n    <NonText_DocumentType_Chars>2955524</NonText_DocumentType_Chars>\n    <HTML_Chars>16751205</HTML_Chars>\n    <XBRL_Chars>3522879</XBRL_Chars>\n    <XML_Chars>6602059</XML_Chars>\n    <N_Exhibits>11</N_Exhibits>\n</FileStats>\n<SEC-Header>\n0000950170-23-001409.hdr.sgml : 20230131\n<ACCEPTANCE-DATETIME>20230130212915\nACCESSION NUMBER:\t\t0000950170-23-001409\nCONFORMED SUBMISSION TYPE:\t10-K\nPUBLIC DOCUMENT COUNT:\t\t128\nCONFORMED PERIOD OF REPORT:\t20221231\nFILED AS OF DATE:\t\t20230131\nDATE AS OF CHANGE:\t\t20230130\n\nFILER:\n\n\tCOMPANY DATA:\t\n\t\tCOMPANY CONFORMED NAME:\t\t\tTesla, Inc.\n\t\tCENTRAL INDEX KEY:\t\t\t0001318605\n\t\tSTANDARD INDUSTRIAL CLASSIFICATION:\tMOTOR VEHICLES & PASSENGER CAR BODIES [3711]\n\t\tIRS NUMBER:\t\t\t\t912197729\n\t\tSTATE OF INCORPORATION:\t\t\tDE\n

#### <span style="color:red">Questions</span>

2. Assume words are those separated by a space ' '. Check how many words are there in the above 10-K?
3. Assume paragraphs are those separated by a double line break '\n\n'. Check how many paragraphs are there in the above 10-K?

#### <span style="color:red">Answer</span>

See the code below

In [15]:
num_words = len(content.split(' '))
num_paras = len(content.split('\n\n'))

print(f'There are {num_words} words and {num_paras} paragraphs in this 10-K file')

There are 129031 words and 2344 paragraphs in this 10-K file


## Read and clean the data

In [16]:
from tqdm.auto import tqdm
from Week_1.Code.Utils import readfile, lowercase, special_character_removal, fix_contraction, stopword_removal, lemmatization

# The text-normalization function
def text_normalization(f):
    text = readfile(f)
    text = lowercase(text)
    text = fix_contraction(text)
    text = special_character_removal(text, keep = '.', remove_number = True)
    text = stopword_removal(text, remove_single_letters = True)
    text = lemmatization(text[:1_000_000])
    return text

# Read, then clean all 10-K files of Tesla
contents = []
for f in tqdm(all_10k_files):
    contents.append(text_normalization(f))

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 345.1/345.1 kB 9.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 114.9/114.9 kB 5.0 MB/s eta 0:00:00


[nltk_data] Downloading package stopwords to /usr/share/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


  0%|          | 0/10 [00:00<?, ?it/s]

## Let's train a Word2Vec model using those clean texts

This can be done using the `gensim` package. First, let's import it by `from gensim.models import Word2Vec`.

An example usage is as follows,

```
model = Word2Vec(
    sentences = corpus      # Your tokenized corpus
    vector_size = 100,      # Dimensionality of the word vectors
    window = 5,             # Context window size
    min_count = 2,          # Ignore words with total frequency < 2
    workers = 4,            # Number of CPU cores to use
    epochs = 5,             # Number of iterations the model goes over the corpus
    sg = 1                  # 1 for Skip-gram; 0 for CBOW
    
)
```


#### 1. Prepare the corpus

The `corpus` argument should look like

```
corpus = [
    ['word_1,1', 'word_1,2', ..., 'word_1,1t'],
    ['word_2,1', 'word_2,2', ..., 'word_2,2t'],
    ...,
    ['word_N,1', 'word_N,2', ..., 'word_N,Nt'],
]
```

With our data, `contents` contains 10-K files. We try to split each of them into sentences first.

In [17]:
from nltk import sent_tokenize

corpus = []

for content in tqdm(contents):
    sentences = sent_tokenize(content)
    corpus.extend(sentences)    # NOTICE: use .extend() here instead of .append()

print(f'There are {len(corpus)} sentences in the whole corpus')

  0%|          | 0/10 [00:00<?, ?it/s]

There are 45524 sentences in the whole corpus


After that, we have to tokenize sentences

In [18]:
from nltk import word_tokenize

tokenized_corpus = []

for sent in tqdm(corpus):
    tokenized_corpus.append(word_tokenize(sent))

  0%|          | 0/45524 [00:00<?, ?it/s]

#### Training

In [19]:
from gensim.models import Word2Vec

word2vec_model = Word2Vec(
    sentences = tokenized_corpus,    # Put the prepared corpus here
    vector_size = 100,
    window = 5,
    min_count = 2,
    workers = 4,
    sg = 1,
    epochs = 30
)

## Evaluating the results

* First, check the vocabulary

In [20]:
print(f'There are {len(word2vec_model.wv)} words in the vocabulary')

There are 6144 words in the vocabulary


* Check the vector of a single words

In [21]:
word = 'profit'
if word in word2vec_model.wv.key_to_index:
    print(f'The embedding of "{word}" is:')
    print(word2vec_model.wv['profit'])
else:
    print('Your chosen word does not appear in the vocabulary, please choose another word!')

The embedding of "profit" is:
[ 0.52409315 -0.05661821  0.79270124  0.851896    0.08860404 -0.4993315
 -0.21500213  0.5228896   0.6858653  -0.54560375  0.27887455 -0.21889599
 -0.1416359   0.80119073 -0.93937683 -0.4146121  -0.27481318 -0.47606003
  0.5331388  -0.6863642  -0.04974891 -0.38073075  0.03857679  0.02377227
 -1.3744779   0.07419452  0.6631895  -0.5309829   0.10716615 -1.0318418
  0.3481544   0.45647928  0.8140142  -0.35623983 -0.15123567  0.3173628
 -0.6783517  -1.0317662   0.47394747 -0.57717556  0.6607662   0.2909286
 -0.73596984 -0.8409292  -0.12209304  0.64162445  0.27567506  1.1627469
 -0.0609083   0.5460107   0.46211916  0.17146528 -0.3400463   0.32738376
  0.07617685  0.90580064 -0.70475143  0.6785915  -0.19973399  0.43838564
  0.06412043 -0.19169638  0.4733894  -0.19423155  0.4788291   0.4477494
  0.630342   -0.1777574   0.19559261  0.5481838  -0.28138384 -0.4480565
 -0.61370754 -0.24608082 -0.6505212   0.3167106  -0.01769876 -0.43879053
 -0.05258877  0.72759765  0.

In [22]:
word2vec_model.wv.most_similar('profit')

[('shifting', 0.7101398706436157),
 ('erosion', 0.6080153584480286),
 ('gross', 0.6010890603065491),
 ('substantive', 0.5622541308403015),
 ('denominated', 0.5549935102462769),
 ('revenues', 0.5471616983413696),
 ('chargeable', 0.5388819575309753),
 ('imprisonment', 0.5375240445137024),
 ('income', 0.5181930065155029),
 ('segment', 0.504891037940979)]

#### <span style="color:red">Questions</span>

2. Evaluate the result, what do you think about our own Word2Vec model?

#### <span style="color:red">Answer</span>

The result doesn't make much sense, with words like `shifting` and `erosion` appears as the most similar words. It might be because the Word2Vec model need to be trained with a large amount of data, while we have only 10 Tesla's 10-K documents.

One way to improve our Word2Vec is to utilize the power of the Google Word2Vec model and adapt it to our data, which we do below

## Fine-tuning a Word2Vec model

We can adapt the pre-trained Google Word2Vec model to our data (i.e., the Tesla's 10-K filings)

This method is called **fine-tuning**. For Word2Vec, it can be done using the following steps

1. Create a new Word2Vec model with same dimensions

In [23]:
finetuned_word2vec_model = Word2Vec(
    vector_size = google_wv.vector_size, 
    min_count = 2,
    sg = 1
)

2. Build the vocabulary using our data, which is `tokenized_corpus`

In [24]:
finetuned_word2vec_model.build_vocab(tokenized_corpus)

3. (**Important**) Inject the pre-trained word vectors, which is `google_wv` in our case

In [25]:
finetuned_word2vec_model.wv.vectors_lockf = np.ones(len(google_wv))    # Enable training
finetuned_word2vec_model.wv.intersect_word2vec_format('/kaggle/input/datasets/sugataghosh/google-word2vec/GoogleNews-vectors-negative300.bin', binary = True)

4. Train the model

In [26]:
finetuned_word2vec_model.train(tokenized_corpus, total_examples = len(tokenized_corpus), epochs = 5)

(3093174, 3521430)

#### Evaluate the fine-tuned model

* The fine-tuned model, combining the knowledge of the Google Word2Vec model and our specific data

In [27]:
finetuned_word2vec_model.wv.most_similar('profit', topn = 5)

[('revenue', 0.5829346179962158),
 ('pretax', 0.5686764717102051),
 ('quarterly', 0.5639331936836243),
 ('revenues', 0.5344869494438171),
 ('income', 0.5063909888267517)]

In [28]:
google_wv.most_similar('profit', topn = 5)

[('profits', 0.802002489566803),
 ('proft', 0.7426661252975464),
 ('pretax_profit', 0.6691470742225647),
 ('pretax_profits', 0.643925130367279),
 ('Profit', 0.6339423060417175)]

In [29]:
word2vec_model.wv.most_similar('profit', topn = 5)

[('shifting', 0.7101398706436157),
 ('erosion', 0.6080153584480286),
 ('gross', 0.6010890603065491),
 ('substantive', 0.5622541308403015),
 ('denominated', 0.5549935102462769)]

# Training a Doc2Vec model

To train a Doc2Vec model, we need to prepare the documents, which is more complicated than Word2Vec.

Training a Doc2Vec model includes the following steps:

1. Prepare the corpus (documents) to have the following form
```
corpus = [
    ['word_1,1', 'word_1,2', ..., 'word_1,1t'],
    ['word_2,1', 'word_2,2', ..., 'word_2,2t'],
    ...,
    ['word_N,1', 'word_N,2', ..., 'word_N,Nt'],
]
```

This is our `tokenized_corpus` as above

2. Tag the document using the `TaggedDocument` method as follows:

```
from gensim.models.doc2vec import TaggedDocument

tagged_documents = [TaggedDocument(words = words, tags = [str(i)]) for i, words in enumerate(corpus)]
```

3. Initialize the model

```
model = Doc2Vec(
    vector_size = 100,   # Dimensionality of the word vectors
    window = 5,          # Context window size
    min_count = 2,       # Ignore words with total frequency < 2
    workers = 4,         # Number of CPU cores to use in parallel
    epochs = 40,         # Number of iterations the model goes over the corpus
    dm = 1               # 1 for Distributed Memory; 0 for DBOW
)
```

4. Build the vocabulary

```
model.build_vocab(tagged_documents)
```

5. Train

```
model.train(tagged_docs)
```

**Let's put everything together!**

## Train a Doc2Vec model

In [30]:
from gensim.models.doc2vec import Doc2Vec, TaggedDocument

tagged_documents = [TaggedDocument(words = words, tags = [str(i)]) for i, words in enumerate(tokenized_corpus)]

doc2vec_model = Doc2Vec(
    vector_size = 100,
    window = 5,
    min_count = 2,
    workers = 4,
    epochs = 40,
    dm = 0
)

doc2vec_model.build_vocab(tagged_documents)

doc2vec_model.train(tagged_documents, total_examples = doc2vec_model.corpus_count, epochs = doc2vec_model.epochs)

If you want to save the model for later use, type `model.save(os.path.join('<path_to_output>', 'doc2vec.model'))`. 

To load a saved model, type `model = Doc2Vec.load(os.path.join('<path_to_output>', 'doc2vec.model'))`

## Post-training, infer the vector of a document

After training a Doc2Vec model, we will use it to infer document embedding using the `model.infer_vector` method. This method takes a **tokenized document as the input**.

An example for a sample text is:

In [31]:
sample_text = tokenized_corpus[100]
print('The sample text is:\n', ' '.join(sample_text))
sentence_vector = doc2vec_model.infer_vector(sample_text)

top_similar_sentences = doc2vec_model.dv.most_similar([sentence_vector], topn = 5)
top_similar_sentences

The sample text is:
 energy generation storage expertise electrical mechanical civil software engineering allow we design engineer manufacture install energy generating storage product component include residential utility scale .


[('100', 0.9665006399154663),
 ('26238', 0.9664967656135559),
 ('14326', 0.9645813703536987),
 ('19164', 0.9493767619132996),
 ('16577', 0.9320551156997681)]

In [32]:
tagged_documents[int(top_similar_sentences[0][0])]

TaggedDocument(words=['energy', 'generation', 'storage', 'expertise', 'electrical', 'mechanical', 'civil', 'software', 'engineering', 'allow', 'we', 'design', 'engineer', 'manufacture', 'install', 'energy', 'generating', 'storage', 'product', 'component', 'include', 'residential', 'utility', 'scale', '.'], tags=['100'])

## Fine-tuning a Doc2Vec model

We can also fine-tune a Doc2Vec model as we did with Word2Vec. However, we can only initialize the word vectors

In [33]:
finetuned_doc2vec_model = Doc2Vec(
    vector_size = google_wv.vector_size,
    window = 5,
    min_count = 2,
    workers = 4,
    dm = 0
)

finetuned_doc2vec_model.build_vocab(tagged_documents)

finetuned_doc2vec_model.wv.vectors_lockf = np.ones(len(finetuned_doc2vec_model.wv))
finetuned_doc2vec_model.wv.intersect_word2vec_format('/kaggle/input/datasets/sugataghosh/google-word2vec/GoogleNews-vectors-negative300.bin', binary = True)

finetuned_doc2vec_model.train(tagged_documents, total_examples = len(tagged_documents), epochs = 5)

In [34]:
sample_text = tokenized_corpus[100]
print('The sample text is:\n', ' '.join(sample_text))
sentence_vector = finetuned_doc2vec_model.infer_vector(sample_text)

top_similar_sentences = finetuned_doc2vec_model.dv.most_similar([sentence_vector], topn = 5)
top_similar_sentences

The sample text is:
 energy generation storage expertise electrical mechanical civil software engineering allow we design engineer manufacture install energy generating storage product component include residential utility scale .


[('26795', 0.9878049492835999),
 ('623', 0.986634373664856),
 ('32359', 0.9862000942230225),
 ('10396', 0.9849621653556824),
 ('14855', 0.9842597842216492)]

In [35]:
' '.join(tokenized_corpus[int(top_similar_sentences[0][0])])

'finally manufacturer installer service provider respect solar generation energy storage system supplier electricity generate store certain solar energy energy storage system install customer provider grid service virtual power plant model impact federal state local regulation policy concern import export component electricity pricing interconnection electricity generation storage equipment electrical grid sale electricity generate third party own system .'

# ELMo

Coding ELMo requires some advanced knowledge about Pytorch, including dataset and dataloader design, neural network construction. So, I decide to only provide the code as reference. Feel free to check the code to train an ELMo model in `Week_3.Code.ELMo` in the lecture's Github repo.